Compilation problem on OFW

```
	APPCHK	build/f7-firmware-C/.extapps/snake_game.fap
	CC	applications_user/shapshup/helpers/shapshup_files.c
	CC	applications_user/shapshup/scenes/shapshup_scene.c
	CC	applications_user/shapshup/scenes/shapshup_scene_load_failed.c
	CC	applications_user/shapshup/scenes/shapshup_scene_load_file.c
	CC	applications_user/shapshup/scenes/shapshup_scene_start.c
	CC	applications_user/shapshup/views/shapshup_main_view.c
	CC	applications_user/shapshup/shapshup.c
applications_user/shapshup/views/shapshup_main_view.c: In function 'shapshup_main_view_draw_scale':
applications_user/shapshup/views/shapshup_main_view.c:484:29: error: 'FontBatteryPercent' undeclared (first use in this function)
  484 |     canvas_set_font(canvas, FontBatteryPercent);
      |                             ^~~~~~~~~~~~~~~~~~
applications_user/shapshup/views/shapshup_main_view.c:484:29: note: each undeclared identifier is reported only once for each function it appears in
	ICONS	build/f7-firmware-C/.extapps/shapshup/shapshup_icons.c
scons: *** [build/f7-firmware-C/.extapps/shapshup/views/shapshup_main_view.o] Error 1

********** FBT ERRORS **********
build/f7-firmware-C/.extapps/shapshup/views/shapshup_main_view.o: Error 1
Error: Process completed with exit code 2.

```


🔗 https://github.com/derskythe/flipperzero-shapshup/actions/runs/6208562495/job/16854877280

--- a/.github/workflows/ofw-build.yml
+++ b/.github/workflows/ofw-build.yml
@@ -1,0 +1,223 @@
+name: "OFW build"
+run-name: "OFW Build ${{ inputs.DEPLOY_TARGET }} by @${{ github.ACTOR }}"
+
+on:
+  workflow_dispatch:
+    inputs:
+      version:
+        description: "Enter version to build or left empty for current version"
+        required: false
+        type: string
+  release:
+    types: [created, prereleased]
+  push:
+    paths:
+      - .github/workflows/build-with-firmware.yml
+
+permissions:
+  contents: write
+  packages: write
+
+concurrency:
+  group: firmware-build-${{ vars.FIRMWARE_VERSION }}-${{ vars.RELEASE_VERSION }}
+  cancel-in-progress: false
+
+jobs:
+  build-and-upload:
+    runs-on: ubuntu-latest
+    env:
+      FAP_INCLUDED_IN_FW: 0
+      REPO_SELF: ${{ vars.REPO_SELF }}
+      OFW_PATH: "applications_user/${{ vars.FAP_NAME }}"
+      RELATIVE_PATH: "applications/external/${{ vars.FAP_NAME }}"
+      CURRENT_VERSION: ${{ vars.RELEASE_VERSION }}
+      RELEASE_VERSION: ${{ vars.RELEASE_VERSION }}
+      ZIP_NAME: ""
+      ZIP_TAG: ""
+      TGZ_NAME: ""
+      TGZ_TAG: ""
+    strategy:
+      fail-fast: false
+      matrix:
+        firmware: [official]
+        include:
+          - firmware: official
+            url: ${{ vars.REPO_OFFICIAL }}
+            version: "official"
+            src-included: 0
+    steps:
+      - name: Set version
+        env:
+          INPUT_VERSION: ${{ inputs.version }}
+          CURRENT_VERSION: ${{ env.CURRENT_VERSION }}
+        shell: pwsh
+        run: |
+          $ReleaseVersion = ([string]::IsNullOrWhitespace($env:INPUT_VERSION) ? $env:CURRENT_VERSION : $env:INPUT_VERSION)
+          Write-Output ('RELEASE_VERSION={0}' -f $ReleaseVersion) >> $env:GITHUB_ENV
+
+      - name: Copy Firmware Files
+        uses: actions/checkout@v3
+        with:
+          repository: "${{ matrix.url }}"
+          clean: "true"
+          submodules: "true"
+          ref: "dev"
+
+      - name: Copy Repo Files
+        if: ${{ matrix.src-included == 0 }}
+        uses: actions/checkout@v3
+        with:
+          repository: "${{ vars.REPO_SELF }}"
+          clean: "true"
+          submodules: "true"
+          path: "${{ env.OFW_PATH }}"
+
+      - name: Print vars about state or repo if Unleashed
+        if: ${{ matrix.src-included == 1 }}
+        shell: pwsh
+        run: |
+          git log --pretty=format:'%s by %C(yellow)%an%Creset (%ad)%n %n%b' --date=local --abbrev-commit --max-count=1
+          if( ${{ env.FAP_INCLUDED_IN_FW }} -eq 1 ) {
+            git submodule set-branch --branch master '${{ env.RELATIVE_PATH }}'
+            git submodule sync '${{ env.RELATIVE_PATH }}'
+          }
+          cd '${{ env.OFW_PATH }}'
+          if ( '${{ env.CURRENT_VERSION }}' -ne '${{ env.RELEASE_VERSION }}' ) {
+            Write-Output '::warning title=Different version::Current version is ${{ env.CURRENT_VERSION }} but we trying to build ${{ env.RELEASE_VERSION }}'
+            git checkout tags/v${{ env.RELEASE_VERSION }} -b tmp-build
+
+            if ( $LASTEXITCODE -ne 0 ) {
+              Write-Error '::error title=Cannot checkout to this version::Error during execution checkout to this tag ${{ env.RELEASE_VERSION }}'
+              exit 1
+            }
+          }
+
+          $Output = (git log --pretty=format:'%s by %C(yellow)%an%Creset (%ad)%n %n%b' --date=local --abbrev-commit --max-count=1)
+
+          if ( $LASTEXITCODE -ne 0 ) {
+            Write-Error '::error title=Invalid checkout::Invalid checkout'
+            exit 1
+          }
+          Write-Output ('::notice title=Git output::{0}' -f $Output)
+
+      - name: Print vars about state or repo if Official
+        if: ${{ matrix.src-included == 0 }}
+        shell: pwsh
+        run: |
+          git log --pretty=format:'%s by %C(yellow)%an%Creset (%ad)%n %n%b' --date=local --abbrev-commit --max-count=1
+          cd '${{ env.OFW_PATH }}'
+
+          if ( '${{ env.CURRENT_VERSION }}' -ne '${{ env.RELEASE_VERSION }}' ) {
+            Write-Output '::warning title=Different version::Current version is ${{ env.CURRENT_VERSION }} but we trying to build ${{ env.RELEASE_VERSION }}'
+            git checkout tags/v${{ env.RELEASE_VERSION }} -b tmp-build
+
+            if ( $LASTEXITCODE -ne 0 ) {
+              Write-Error '::error title=Cannot checkout to this version::Error during execution checkout to this tag ${{ env.RELEASE_VERSION }}'
+              exit 1
+            }
+          } else {
+            $Output = (git log --pretty=format:'%s by %C(yellow)%an%Creset (%ad)%n %n%b' --date=local --abbrev-commit --max-count=1)
+
+            if ( $LASTEXITCODE -ne 0 ) {
+              Write-Error '::error title=Invalid checkout::Invalid checkout'
+              exit 1
+            }
+            Write-Output ('::notice title=Git output::{0}' -f $Output)
+          }
+
+      - name: Remove other apps
+        shell: pwsh
+        if: ${{ success() }}
+        # rm to remove problem FAP which includes non-existent files
+        run: |
+          Remove-Item -Force -Recurse ./applications/debug -ErrorAction SilentlyContinue
+          Remove-Item -Force -Recurse ./applications/examples -ErrorAction SilentlyContinue
+
+      - name: Build Firmware
+        shell: bash
+        if: ${{ success() }}
+        env:
+          FBT_NO_SYNC: 0
+          DIST_SUFFIX: ${{ matrix.version }}
+          WORKFLOW_BRANCH_OR_TAG: release-cfw
+        run: |
+          ./fbt COMPACT=1 DEBUG=0 FBT_NO_SYNC=0
+
+      - name: Build FAPs
+        shell: bash
+        if: ${{ success() }}
+        env:
+          FBT_NO_SYNC: 0
+          DIST_SUFFIX: ${{ matrix.version }}
+          WORKFLOW_BRANCH_OR_TAG: release-cfw
+          # rm to remove problem FAP which includes non-existent files
+        run: |
+          ./fbt COMPACT=1 DEBUG=0 FBT_NO_SYNC=0 fap_dist
+
+      - name: Create assets
+        if: ${{ success() }}
+        shell: pwsh
+        env:
+          ZIP_NAME: "${{ vars.FAP_NAME }}_${{ env.RELEASE_VERSION }}_${{ matrix.firmware }}.zip"
+          TGZ_NAME: "${{ vars.FAP_NAME }}_${{ env.RELEASE_VERSION }}_${{ matrix.firmware }}.tgz"
+        run: |
+          function Format-Bytes {
+            param(
+                  [int]$number
+              )
+              $sizes = 'KB', 'MB', 'GB', 'TB', 'PB'
+              for ($x = 0; $x -lt $sizes.count; $x++) {
+                  if ($number -lt [int64]"1$($sizes[$x])") {
+                      if ($x -eq 0) {
+                          return "$number B"
+                      }
+                      else {
+                          $num = $number / [int64]"1$($sizes[$x-1])"
+                          $num = "{0:N2}" -f $num
+                          return "$num $($sizes[$x-1])"
+                      }
+                  }
+              }
+          }
+          $ZipName = $env:ZIP_NAME
+          $TgzName = $env:TGZ_NAME
+          $FapNamme = '${{ vars.FAP_NAME }}.fap'
+          $DstFap = "./$FapNamme"
+          $AppDir = "dist/f7-C/apps/Sub-GHz"
+
+          if (!(Test-Path -Path "$AppDir/$FapNamme" -PathType Leaf)) {
+              Write-Error '::error title=Files not found::Cannot find files in location'
+              exit 1
+          }
+
+          $Size = (Get-Item -Path "$AppDir/$FapNamme" | Get-ItemPropertyValue -Name Length)
+          Write-Output ('Filesize: {0}' -f (Format-Bytes $Size))
+          Copy-Item -Force -Verbose -Path "$AppDir/$FapNamme" -Destination $DstFap
+
+          zip -r -qq $ZipName $DstFap
+          tar zcf $TgzName $DstFap
+
+          if ( !(Test-Path -Path $ZipName -PathType Leaf) -or !(Test-Path -Path $TgzName -PathType Leaf) ) {
+              Write-Error '::error title=Files not found::Cannot find files in location'
+              exit 1
+          }
+
+          $ZipSize = Format-Bytes (Get-Item -Path $ZipName).Length
+          $TgzSize = Format-Bytes (Get-Item -Path $TgzName ).Length
+
+          Write-Output ('ZIP_NAME={0}' -f $ZipName) >> $env:GITHUB_ENV
+          Write-Output ('TGZ_NAME={0}' -f $TgzName ) >> $env:GITHUB_ENV
+          Write-Output ('ZIP_TAG={0} ({1})' -f $ZipName, $ZipSize) >> $env:GITHUB_ENV
+          Write-Output ('TGZ_TAG={0} ({1})' -f $TgzName , $TgzSize) >> $env:GITHUB_ENV
+
+      - name: Upload assets
+        if: ${{ success() && env.ZIP_NAME != '' }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.FLIPPER_TOKEN }}
+        run: |
+          gh release create v${{ env.RELEASE_VERSION }} --generate-notes --draft -R ${{ env.REPO_SELF }}
+          gh release upload 'v${{ env.RELEASE_VERSION }}' '${{ env.ZIP_NAME }}#${{ env.ZIP_TAG }}' \
+            '${{ env.TGZ_NAME }}#${{ env.TGZ_TAG }}' --clobber -R ${{ env.REPO_SELF }}
+          gh release edit 'v${{ env.RELEASE_VERSION }}' --draft=false -R ${{ env.REPO_SELF }}
+
+#EOF

--- a/helpers/gui_top_buttons.h
+++ b/helpers/gui_top_buttons.h
@@ -7,15 +7,19 @@
 #include <assets_icons.h>
 
 /**
- * Thanks to the author of metronome
+ * @brief draw button in top left corner of screen
+ *
+ * @author panki27 (https://github.com/panki27/Metronome)
  * @param canvas Canvas* 
- * @param str const char*
+ * @param str caption of button
  */
 void elements_button_top_left(Canvas* canvas, const char* str);
 
 /**
- * Thanks to the author of metronome
+ * @brief draw button in top right corner of screen
+ * 
+ * @author panki27 (https://github.com/panki27/Metronome)
  * @param canvas Canvas*
- * @param str const char*
+ * @param str caption of button
  */
 void elements_button_top_right(Canvas* canvas, const char* str);

--- a/helpers/shapshup_files.c
+++ b/helpers/shapshup_files.c
@@ -1,19 +1,4 @@
-#include <stdio.h>
-#include <stddef.h>
-#include <stdbool.h>
 #include <inttypes.h>
-//#include "../../../lib/toolbox/path.h"
-//#include "../../../lib/toolbox/stream/stream.h"
-//#include "../../../furi/core/string.h"
-//#include "../../../furi/furi.h"
-//#include "../../../lib/flipper_format/flipper_format_stream_i.h"
-//#include "../../../applications/services/storage/storage.h"
-//#include "../../../furi/core/check.h"
-//#include "../../../furi/core/record.h"
-//#include "../../../lib/flipper_format/flipper_format.h"
-//#include "../../../furi/core/log.h"
-//#include "../../../lib/flipper_format/flipper_format_i.h"
-//#include "../../../lib/subghz/types.h"
 #include "shapshup_files.h"
 
 #define TAG "ShapShupFiles"

--- a/helpers/shapshup_files.c
+++ b/helpers/shapshup_files.c
@@ -1,5 +1,6 @@
-#include <inttypes.h>
 #include "shapshup_files.h"
+#include <inttypes.h>
+#include <lib/subghz/types.h>
 
 #define TAG "ShapShupFiles"
 #define RAW_KEY_NAME "RAW_Data"

--- a/helpers/shapshup_files.h
+++ b/helpers/shapshup_files.h
@@ -1,11 +1,9 @@
 #pragma once
 #include <furi.h>
+#include "m-array.h"
+#include <toolbox/path.h>
 #include <flipper_format_i.h>
 #include <flipper_format_stream_i.h>
-#include <subghz/types.h>
-#include <toolbox/path.h>
-#include <subghz/types.h>
-#include "m-array.h"
 
 /**
  * @brief file operation result response

--- a/helpers/shapshup_files.h
+++ b/helpers/shapshup_files.h
@@ -7,6 +7,10 @@
 #include <subghz/types.h>
 #include "m-array.h"
 
+/**
+ * @brief file operation result response
+ * 
+ */
 typedef enum {
     ShapShupFileResultOk,
     ShapShupFileResultOpenError,
@@ -21,8 +25,16 @@ typedef enum {
     ShapShupFileResultTotal
 } ShapShupFileResults;
 
+/**
+ * @brief Construct a new array def object
+ * 
+ */
 ARRAY_DEF(array_raw, int32_t)
 
+/**
+ * @brief contains info about .sub file
+ * 
+ */
 typedef struct {
     ShapShupFileResults result;
     uint32_t frequency;
@@ -46,25 +58,24 @@ ShapShupRawFile* load_file_shapshup(const char* file_path);
 /**
  * @brief read_int32_shapshup
  * 
- * @param stream 
- * @param _data 
- * @param data_size 
- * @return true 
- * @return false 
+ * @param stream stream to read
+ * @param _data return readed data
+ * @param data_size size of data to read
+ * @return true/false
  */
 bool read_int32_shapshup(Stream* stream, int32_t* _data, uint16_t data_size);
 
 /**
- * @brief shapshup_files_result_description
+ * @brief get description about file operation
  * 
- * @param index 
- * @return const char* 
+ * @param result_code struct @see ShapShupFileResults
+ * @return human readable description about result code
  */
-const char* shapshup_files_result_description(ShapShupFileResults index);
+const char* shapshup_files_result_description(ShapShupFileResults result_code);
 
 /**
- * @brief clean_raw_values
+ * @brief make clean of array to read new values
  * 
- * @param raw_file 
+ * @param raw_file @see ShapShupRawFile
  */
 void clean_raw_values(ShapShupRawFile* raw_file);

--- a/shapshup.c
+++ b/shapshup.c
@@ -76,6 +76,11 @@ ShapShupState* shapshup_alloc() {
     view_dispatcher_add_view(
         instance->view_dispatcher, ShapShupViewPopup, popup_get_view(instance->popup));
 
+    // Loading
+    instance->loading = loading_alloc();
+    view_dispatcher_add_view(
+        instance->view_dispatcher, ShapShupViewLoading, loading_get_view(instance->loading));
+
     // ViewStack
     instance->view_stack = view_stack_alloc();
     view_dispatcher_add_view(
@@ -121,6 +126,10 @@ void shapshup_free(ShapShupState* instance) {
     view_dispatcher_remove_view(instance->view_dispatcher, ShapShupViewPopup);
     popup_free(instance->popup);
 
+    // Loading
+    view_dispatcher_remove_view(instance->view_dispatcher, ShapShupViewLoading);
+    loading_free(instance->loading);
+
     // ViewStack
     view_dispatcher_remove_view(instance->view_dispatcher, ShapShupViewStack);
     view_stack_free(instance->view_stack);
@@ -150,19 +159,16 @@ void shapshup_free(ShapShupState* instance) {
 }
 
 void shapshup_show_loading_popup(void* context, bool show) {
-    TaskHandle_t timer_task = xTaskGetHandle(configTIMER_SERVICE_TASK_NAME);
+    furi_assert(context);
     ShapShupState* instance = context;
-    ViewStack* view_stack = instance->view_stack;
-    Loading* loading = instance->loading;
 
     if(show) {
         // Raise timer priority so that animations can play
-        vTaskPrioritySet(timer_task, configMAX_PRIORITIES - 1);
-        view_stack_add_view(view_stack, loading_get_view(loading));
+        furi_timer_set_thread_priority(FuriTimerThreadPriorityElevated);
+        view_dispatcher_switch_to_view(instance->view_dispatcher, ShapShupViewPopup);
     } else {
-        view_stack_remove_view(view_stack, loading_get_view(loading));
         // Restore default timer priority
-        vTaskPrioritySet(timer_task, configTIMER_TASK_PRIORITY);
+        furi_timer_set_thread_priority(FuriTimerThreadPriorityNormal);
     }
 }
 
@@ -181,7 +187,7 @@ void shapshup_popup_closed_callback(void* context) {
 }
 
 // ENTRYPOINT
-__attribute__((unused)) int32_t shapshup_app(void* p) {
+int32_t shapshup_app(void* p) {
     UNUSED(p);
 
     ShapShupState* instance = shapshup_alloc();

--- a/shapshup.c
+++ b/shapshup.c
@@ -14,7 +14,7 @@
 #include "shapshup_custom_event.h"
 
 #define TAG "ShapShupApp"
-#define DELAY_MS 100
+#define TICK_PERIOD 500
 
 static bool shapshup_custom_event_callback(void* context, uint32_t event) {
     furi_assert(context);
@@ -51,7 +51,7 @@ ShapShupState* shapshup_alloc() {
     view_dispatcher_set_navigation_event_callback(
         instance->view_dispatcher, shapshup_back_event_callback);
     view_dispatcher_set_tick_event_callback(
-        instance->view_dispatcher, shapshup_tick_event_callback, 500);
+        instance->view_dispatcher, shapshup_tick_event_callback, TICK_PERIOD);
 
     //Dialog
     instance->dialogs = furi_record_open(RECORD_DIALOGS);
@@ -165,7 +165,7 @@ void shapshup_show_loading_popup(void* context, bool show) {
     if(show) {
         // Raise timer priority so that animations can play
         furi_timer_set_thread_priority(FuriTimerThreadPriorityElevated);
-        view_dispatcher_switch_to_view(instance->view_dispatcher, ShapShupViewPopup);
+        view_dispatcher_switch_to_view(instance->view_dispatcher, ShapShupViewLoading);
     } else {
         // Restore default timer priority
         furi_timer_set_thread_priority(FuriTimerThreadPriorityNormal);
@@ -186,7 +186,12 @@ void shapshup_popup_closed_callback(void* context) {
         instance->view_dispatcher, ShapShupCustomEventTypePopupClosed);
 }
 
-// ENTRYPOINT
+/**
+ * @brief Entrypoint
+ * 
+ * @param p 
+ * @return int32_t 
+ */
 int32_t shapshup_app(void* p) {
     UNUSED(p);
 

--- a/shapshup.h
+++ b/shapshup.h
@@ -2,6 +2,6 @@
 
 /**
  * @brief ShapShupState
- * 
+ * See: @struct ShapShupState
  */
 typedef struct ShapShupState ShapShupState;

--- a/shapshup_custom_event.h
+++ b/shapshup_custom_event.h
@@ -1,8 +1,5 @@
 #pragma once
 
-#include <stdint.h>
-#include <stddef.h>
-
 /**
  * @brief ShapshupCustomEvent
 *  ShapShupCustomEventTypeReserved = 100,

--- a/shapshup_i.h
+++ b/shapshup_i.h
@@ -34,17 +34,40 @@
 #include "views/shapshup_main_view.h"
 #include "helpers/shapshup_files.h"
 
-#define SHAPSHUP_TEXT_STORE_SIZE 256
-
+/**
+ * @brief max size of text field
+ * 
+ */
+#define SHAPSHUP_TEXT_STORE_SIZE 255
+/**
+ * @brief max size of filename
+ * 
+ */
 #define SHAPSHUP_MAX_LEN_NAME 64
+/**
+ * @brief default directory to show
+ * 
+ */
 #define SHAPSHUP_PATH EXT_PATH("subghz")
+/**
+ * @brief filename extension to show
+ * 
+ */
 #define SHAPSHUP_FILE_EXT ".sub"
+
+/**
+ * @brief used for DEBUG purpose
+ * 
+ */
 //#define SHAPSHUP_FAST_TRACK false
 
+/**
+ * @brief view mode
+ * 
+ */
 typedef enum {
     ShapShupViewNone,
     ShapShupViewMain,
-    ShapShupViewAttack,
     ShapShupViewTextInput,
     ShapShupViewDialogEx,
     ShapShupViewPopup,
@@ -53,6 +76,10 @@ typedef enum {
     ShapShupViewStack,
 } ShapShupView;
 
+/**
+ * @brief main state of app
+ * 
+ */
 struct ShapShupState {
     // GUI elements
     NotificationApp* notifications;
@@ -65,19 +92,13 @@ struct ShapShupState {
     DialogsApp* dialogs;
     Loading* loading;
 
-    // Other
     char text_store[SHAPSHUP_MAX_LEN_NAME];
     FuriString* file_path;
 
-    // Views
-    ShapShupMainView* view_main;
-    ShapShupView current_view;
-
-    // Scene
-    SceneManager* scene_manager;
-
-    // Storage
-    Storage* storage;
+    ShapShupMainView* view_main; /** View main, default view*/
+    ShapShupView current_view; /** @enum ShapShupView for track enums */
+    SceneManager* scene_manager; /** Scene manager*/
+    Storage* storage; /** Storage pointer */
 };
 
 /**

--- a/shapshup_i.h
+++ b/shapshup_i.h
@@ -9,9 +9,6 @@
 #include <notification/notification_messages.h>
 #include <string.h>
 
-#include <lib/toolbox/stream/stream.h>
-#include <stream_buffer.h>
-
 #include <gui/gui.h>
 #include <gui/view_dispatcher.h>
 #include <gui/view_stack.h>
@@ -29,6 +26,8 @@
 #include <lib/subghz/environment.h>
 #include <notification/notification.h>
 #include <notification/notification_messages.h>
+
+#include <lib/toolbox/stream/stream.h>
 
 #include "shapshup.h"
 #include "scenes/shapshup_scene.h"
@@ -49,6 +48,7 @@ typedef enum {
     ShapShupViewTextInput,
     ShapShupViewDialogEx,
     ShapShupViewPopup,
+    ShapShupViewLoading,
     ShapShupViewWidget,
     ShapShupViewStack,
 } ShapShupView;

--- a/views/shapshup_main_view.c
+++ b/views/shapshup_main_view.c
@@ -135,144 +135,144 @@ bool shapshup_main_view_input(InputEvent* event, void* context) {
     ShapShupMainView* instance = context;
 
     if(event->key == InputKeyBack && event->type == InputTypeLong) {
-        // Consumed
-        return false;
+    // Consumed
+    return false;
     } else if(event->key == InputKeyBack) {
-        shapshup_main_view_show_alert(instance, "Press long BACK to exit", 150);
-        return true;
+    shapshup_main_view_show_alert(instance, "Press long BACK to EXIT", SHAPSHUP_ALERT_DEFAULT_TTL);
+    return true;
     }
 
     bool is_clicked = (event->type == InputTypeShort) || (event->type == InputTypeRepeat);
 
     if(is_clicked) {
-        uint64_t offset_per_page =
+    uint64_t offset_per_page =
+        instance->raw_file != NULL ?
+            calc_offset_per_page(
+                instance->raw_file->total_len, instance->raw_file->min_len, instance->scale) :
+            0;
+    if(event->key == InputKeyOk) {
+        shapshup_main_view_reset_alert(instance);
+        instance->callback(ShapShupCustomEventTypeLoadFile, instance->context);
+    } else if(event->key == InputKeyLeft && instance->offset > 0) {
+        shapshup_main_view_reset_alert(instance);
+        instance->offset = instance->offset < offset_per_page ? 0 :
+                                                                instance->offset - offset_per_page;
+#ifdef FURI_DEBUG
+        FURI_LOG_D(
+            TAG,
+            "L offset: %lld, offset_per_page: %lld, offset: %lld, total_len: %lld",
+            instance->offset,
+            offset_per_page,
+            instance->offset,
+            instance->raw_file->total_len);
+#endif
+        shapshup_main_view_array_reset(instance);
+        shapshup_main_view_create_shapes(instance, offset_per_page);
+
+        with_view_model(
+            instance->view,
+            ShapShupMainViewModel * model,
+            {
+                model->shape_list = instance->shape_list;
+                model->count_shapes = instance->count_shapes;
+                model->offset = instance->offset;
+                model->offset_per_page = offset_per_page;
+            },
+            true);
+    } else if(event->key == InputKeyRight) {
+        shapshup_main_view_reset_alert(instance);
+        uint64_t calc = instance->offset + offset_per_page;
+#ifdef FURI_DEBUG
+        FURI_LOG_D(
+            TAG,
+            "R offset: %lld, offset_per_page: %lld, calc: %lld, total_len: %lld",
+            instance->offset,
+            offset_per_page,
+            calc,
+            instance->raw_file->total_len);
+#endif
+        if(calc < instance->raw_file->total_len) {
+            instance->offset = calc;
+        } else if(
+            calc > instance->raw_file->total_len &&
+            instance->raw_file->total_len - calc < offset_per_page) {
+            instance->offset = instance->offset > 0 ? calc : 0;
+        }
+        shapshup_main_view_array_reset(instance);
+        shapshup_main_view_create_shapes(instance, offset_per_page);
+
+        with_view_model(
+            instance->view,
+            ShapShupMainViewModel * model,
+            {
+                model->shape_list = instance->shape_list;
+                model->count_shapes = instance->count_shapes;
+                model->offset = instance->offset;
+                model->offset_per_page = offset_per_page;
+            },
+            true);
+    } else if(event->key == InputKeyDown) {
+        instance->scale -= SCALE_STEP;
+        if(instance->scale < SCALE_STEP) {
+            instance->scale = SHAPSHUP_SCALE_MIN_VALUE;
+        }
+        offset_per_page =
             instance->raw_file != NULL ?
                 calc_offset_per_page(
                     instance->raw_file->total_len, instance->raw_file->min_len, instance->scale) :
                 0;
-        if(event->key == InputKeyOk) {
-            shapshup_main_view_reset_alert(instance);
-            instance->callback(ShapShupCustomEventTypeLoadFile, instance->context);
-        } else if(event->key == InputKeyLeft && instance->offset > 0) {
-            shapshup_main_view_reset_alert(instance);
-            instance->offset =
-                instance->offset < offset_per_page ? 0 : instance->offset - offset_per_page;
 #ifdef FURI_DEBUG
-            FURI_LOG_D(
-                TAG,
-                "L offset: %lld, offset_per_page: %lld, offset: %lld, total_len: %lld",
-                instance->offset,
-                offset_per_page,
-                instance->offset,
-                instance->raw_file->total_len);
+        FURI_LOG_D(TAG, "Scale: %f", (double)instance->scale);
 #endif
-            shapshup_main_view_array_reset(instance);
-            shapshup_main_view_create_shapes(instance, offset_per_page);
+        shapshup_main_view_array_reset(instance);
+        shapshup_main_view_create_shapes(instance, offset_per_page);
 
-            with_view_model(
-                instance->view,
-                ShapShupMainViewModel * model,
-                {
-                    model->shape_list = instance->shape_list;
-                    model->count_shapes = instance->count_shapes;
-                    model->offset = instance->offset;
-                    model->offset_per_page = offset_per_page;
-                },
-                true);
-        } else if(event->key == InputKeyRight) {
-            shapshup_main_view_reset_alert(instance);
-            uint64_t calc = instance->offset + offset_per_page;
-#ifdef FURI_DEBUG
-            FURI_LOG_D(
-                TAG,
-                "R offset: %lld, offset_per_page: %lld, calc: %lld, total_len: %lld",
-                instance->offset,
-                offset_per_page,
-                calc,
-                instance->raw_file->total_len);
-#endif
-            if(calc < instance->raw_file->total_len) {
-                instance->offset = calc;
-            } else if(
-                calc > instance->raw_file->total_len &&
-                instance->raw_file->total_len - calc < offset_per_page) {
-                instance->offset = instance->offset > 0 ? calc : 0;
-            }
-            shapshup_main_view_array_reset(instance);
-            shapshup_main_view_create_shapes(instance, offset_per_page);
-
-            with_view_model(
-                instance->view,
-                ShapShupMainViewModel * model,
-                {
-                    model->shape_list = instance->shape_list;
-                    model->count_shapes = instance->count_shapes;
-                    model->offset = instance->offset;
-                    model->offset_per_page = offset_per_page;
-                },
-                true);
-        } else if(event->key == InputKeyDown) {
-            instance->scale -= SCALE_STEP;
-            if(instance->scale < SCALE_STEP) {
-                instance->scale = 0.1f;
-            }
-            offset_per_page = instance->raw_file != NULL ? calc_offset_per_page(
-                                                               instance->raw_file->total_len,
-                                                               instance->raw_file->min_len,
-                                                               instance->scale) :
-                                                           0;
-#ifdef FURI_DEBUG
-            FURI_LOG_D(TAG, "Scale: %f", (double)instance->scale);
-#endif
-            shapshup_main_view_array_reset(instance);
-            shapshup_main_view_create_shapes(instance, offset_per_page);
-
-            shapshup_main_view_show_alert(instance, "Zoom-in", 150);
-            with_view_model(
-                instance->view,
-                ShapShupMainViewModel * model,
-                {
-                    model->shape_list = instance->shape_list;
-                    model->count_shapes = instance->count_shapes;
-                    model->scale = instance->scale;
-                    model->offset_per_page = offset_per_page;
-                },
-                true);
-        } else if(event->key == InputKeyUp) {
-            if(instance->scale < SCALE_STEP) {
-                instance->scale = SCALE_STEP;
-            } else if(SHAPSHUP_DEFAULT_SCALE_STEP - instance->scale > 0.01f) {
-                instance->scale += SCALE_STEP;
-            }
-
-            if(instance->scale > SHAPSHUP_DEFAULT_SCALE_STEP) {
-                instance->scale = SHAPSHUP_DEFAULT_SCALE_STEP;
-            }
-            offset_per_page = instance->raw_file != NULL ? calc_offset_per_page(
-                                                               instance->raw_file->total_len,
-                                                               instance->raw_file->min_len,
-                                                               instance->scale) :
-                                                           0;
-#ifdef FURI_DEBUG
-            FURI_LOG_D(TAG, "Scale: %f", (double)instance->scale);
-#endif
-            shapshup_main_view_array_reset(instance);
-            shapshup_main_view_create_shapes(instance, offset_per_page);
-
-            shapshup_main_view_show_alert(instance, "Zoom-out", 150);
-            with_view_model(
-                instance->view,
-                ShapShupMainViewModel * model,
-                {
-                    model->shape_list = instance->shape_list;
-                    model->count_shapes = instance->count_shapes;
-                    model->scale = instance->scale;
-                    model->offset_per_page = offset_per_page;
-                },
-                true);
+        shapshup_main_view_show_alert(instance, "Zoom-in", SHAPSHUP_ALERT_DEFAULT_TTL);
+        with_view_model(
+            instance->view,
+            ShapShupMainViewModel * model,
+            {
+                model->shape_list = instance->shape_list;
+                model->count_shapes = instance->count_shapes;
+                model->scale = instance->scale;
+                model->offset_per_page = offset_per_page;
+            },
+            true);
+    } else if(event->key == InputKeyUp) {
+        if(instance->scale < SCALE_STEP) {
+            instance->scale = SCALE_STEP;
+        } else if(SHAPSHUP_DEFAULT_SCALE_STEP - instance->scale > SHAPSHUP_SCALE_MIN_VALUE) {
+            instance->scale += SCALE_STEP;
         }
 
-        consumed = true;
+        if(instance->scale > SHAPSHUP_DEFAULT_SCALE_STEP) {
+            instance->scale = SHAPSHUP_DEFAULT_SCALE_STEP;
+        }
+        offset_per_page =
+            instance->raw_file != NULL ?
+                calc_offset_per_page(
+                    instance->raw_file->total_len, instance->raw_file->min_len, instance->scale) :
+                0;
+#ifdef FURI_DEBUG
+        FURI_LOG_D(TAG, "Scale: %f", (double)instance->scale);
+#endif
+        shapshup_main_view_array_reset(instance);
+        shapshup_main_view_create_shapes(instance, offset_per_page);
+
+        shapshup_main_view_show_alert(instance, "Zoom-out", SHAPSHUP_ALERT_DEFAULT_TTL);
+        with_view_model(
+            instance->view,
+            ShapShupMainViewModel * model,
+            {
+                model->shape_list = instance->shape_list;
+                model->count_shapes = instance->count_shapes;
+                model->scale = instance->scale;
+                model->offset_per_page = offset_per_page;
+            },
+            true);
+    }
+
+    consumed = true;
     }
 
     return consumed;
@@ -350,39 +350,39 @@ View* shapshup_main_view_get_view(ShapShupMainView* instance) {
 ShapShupFileResults shapshup_main_view_load_file(ShapShupMainView* instance, const char* name) {
     furi_assert(instance);
     if(instance->raw_file != NULL) {
-        free(instance->raw_file);
+    free(instance->raw_file);
     }
     instance->raw_file = load_file_shapshup(name);
 
     if(instance->raw_file->result == ShapShupFileResultOk) {
-        instance->offset = 0;
-        uint64_t offset_per_page = calc_offset_per_page(
-            instance->raw_file->total_len, instance->raw_file->min_len, instance->scale);
+    instance->offset = 0;
+    uint64_t offset_per_page = calc_offset_per_page(
+        instance->raw_file->total_len, instance->raw_file->min_len, instance->scale);
 
-        // Reset array
-        shapshup_main_view_array_reset(instance);
-        shapshup_main_view_create_shapes(instance, offset_per_page);
+    // Reset an array
+    shapshup_main_view_array_reset(instance);
+    shapshup_main_view_create_shapes(instance, offset_per_page);
 
-        with_view_model(
-            instance->view,
-            ShapShupMainViewModel * model,
-            {
-                model->raw_file = instance->raw_file;
-                model->shape_list = instance->shape_list;
-                model->count_shapes = instance->count_shapes;
-                model->offset = 0;
-                if(model->raw_file != NULL) {
-                    model->offset_per_page = offset_per_page;
-                    if(model->raw_file->total_len > 1000) {
-                        model->calc_total_len = (uint64_t)model->raw_file->total_len / 1000;
-                        model->is_ms = true;
-                    } else {
-                        model->is_ms = false;
-                    }
+    with_view_model(
+        instance->view,
+        ShapShupMainViewModel * model,
+        {
+            model->raw_file = instance->raw_file;
+            model->shape_list = instance->shape_list;
+            model->count_shapes = instance->count_shapes;
+            model->offset = 0;
+            if(model->raw_file != NULL) {
+                model->offset_per_page = offset_per_page;
+                if(model->raw_file->total_len > 1000) {
+                    model->calc_total_len = (uint64_t)model->raw_file->total_len / 1000;
+                    model->is_ms = true;
+                } else {
+                    model->is_ms = false;
                 }
-                model->scale = SHAPSHUP_DEFAULT_SCALE_STEP;
-            },
-            true);
+            }
+            model->scale = SHAPSHUP_DEFAULT_SCALE_STEP;
+        },
+        true);
     }
 
     return instance->raw_file->result;
@@ -404,76 +404,76 @@ void shapshup_main_view_create_shapes(ShapShupMainView* instance, uint64_t offse
     //     chunk);
 #endif
     for(uint64_t i = 0; i < instance->raw_file->total_count && !last; i++) {
-        int32_t value = *array_raw_get(instance->raw_file->values, i);
+    int32_t value = *array_raw_get(instance->raw_file->values, i);
 #ifdef FURI_DEBUG
-        //FURI_LOG_D(TAG, "value: %ld, step: %lld", value, i);
+    //FURI_LOG_D(TAG, "value: %ld, step: %lld", value, i);
 #endif
-        uint64_t current_value;
+    uint64_t current_value;
 
-        if(value < 0) {
-            current_value = value * -1;
-            is_negative = true;
-        } else {
-            current_value = value;
-            is_negative = false;
-        }
-        if(current_offset + current_value < instance->offset) {
-            current_offset += current_value;
-#ifdef FURI_DEBUG
-            // FURI_LOG_I(
-            //     TAG,
-            //     "continue current_offset: %lld, current_value: %lld",
-            //     current_offset,
-            //     current_value);
-#endif
-            continue;
-            /*} else if(model->offset > current_offset) {
-                skip_value = model->offset - current_offset;*/
-        } else {
-            skip_value = 0;
-        }
-
-        current_value = skip_value + current_value;
-#ifdef FURI_DEBUG
-        // FURI_LOG_D(TAG, "skip_value: %lld, current_value: %lld", skip_value, current_value);
-#endif
-        if(current_value > offset_per_page) {
-            last = true;
-            current_value = current_value - (current_value - offset_per_page);
-            // FURI_LOG_I(
-            //     TAG,
-            //     "current_value > model->offset_per_page: %lld, model->offset_per_page: %lld",
-            //     current_value,
-            //     model->offset_per_page);
-        } else if(current_offset + current_value > instance->offset + offset_per_page) {
-            last = true;
-            current_value = current_value - ((current_offset + current_value) -
-                                             (instance->offset + offset_per_page));
-            // FURI_LOG_I(
-            //     TAG,
-            //     "current_offset + current_value > current_offset + model->offset_per_page: %lld, model->offset_per_page: %lld",
-            //     current_value,
-            //     model->offset_per_page);
-        }
+    if(value < 0) {
+        current_value = value * -1;
+        is_negative = true;
+    } else {
+        current_value = value;
+        is_negative = false;
+    }
+    if(current_offset + current_value < instance->offset) {
         current_offset += current_value;
-        current_value = (uint64_t)(current_value / chunk);
+#ifdef FURI_DEBUG
+        // FURI_LOG_I(
+        //     TAG,
+        //     "continue current_offset: %lld, current_value: %lld",
+        //     current_offset,
+        //     current_value);
+#endif
+        continue;
+        /*} else if(model->offset > current_offset) {
+                skip_value = model->offset - current_offset;*/
+    } else {
+        skip_value = 0;
+    }
 
-        ShapShupShapeItem* item = ShapShupShapeItemArray_push_raw(instance->shape_list->data);
-        if(is_negative) {
-            item->x = current_x;
-            item->y = SHAPSHUP_CHART_LOWEST_POINT - 2;
-            item->width = current_value + 1;
-            item->height = 2;
-        } else {
-            item->x = current_x;
-            item->y = SHAPSHUP_CHART_HIGHEST_POINT;
-            item->width = current_value + 1;
-            item->height = SHAPSHUP_CHART_LOWEST_POINT - SHAPSHUP_CHART_HIGHEST_POINT;
-        }
+    current_value = skip_value + current_value;
+#ifdef FURI_DEBUG
+    // FURI_LOG_D(TAG, "skip_value: %lld, current_value: %lld", skip_value, current_value);
+#endif
+    if(current_value > offset_per_page) {
+        last = true;
+        current_value = current_value - (current_value - offset_per_page);
+        // FURI_LOG_I(
+        //     TAG,
+        //     "current_value > model->offset_per_page: %lld, model->offset_per_page: %lld",
+        //     current_value,
+        //     model->offset_per_page);
+    } else if(current_offset + current_value > instance->offset + offset_per_page) {
+        last = true;
+        current_value = current_value -
+                        ((current_offset + current_value) - (instance->offset + offset_per_page));
+        // FURI_LOG_I(
+        //     TAG,
+        //     "current_offset + current_value > current_offset + model->offset_per_page: %lld, model->offset_per_page: %lld",
+        //     current_value,
+        //     model->offset_per_page);
+    }
+    current_offset += current_value;
+    current_value = (uint64_t)(current_value / chunk);
 
-        current_x += current_value == 0 ? 1 : current_value;
-        current_x++;
-        instance->count_shapes++;
+    ShapShupShapeItem* item = ShapShupShapeItemArray_push_raw(instance->shape_list->data);
+    if(is_negative) {
+        item->x = current_x;
+        item->y = SHAPSHUP_CHART_LOWEST_POINT - 2;
+        item->width = current_value + 1;
+        item->height = 2;
+    } else {
+        item->x = current_x;
+        item->y = SHAPSHUP_CHART_HIGHEST_POINT;
+        item->width = current_value + 1;
+        item->height = SHAPSHUP_CHART_LOWEST_POINT - SHAPSHUP_CHART_HIGHEST_POINT;
+    }
+
+    current_x += current_value == 0 ? 1 : current_value;
+    current_x++;
+    instance->count_shapes++;
     }
 }
 
@@ -488,9 +488,9 @@ void shapshup_main_view_draw_scale(Canvas* canvas, ShapShupMainViewModel* model)
     canvas_draw_line(canvas, 1, SHAPSHUP_TOP_SCALE, 1, SHAPSHUP_TOP_SCALE + 4);
     canvas_draw_line(canvas, width - 1, SHAPSHUP_TOP_SCALE, width - 1, SHAPSHUP_TOP_SCALE + 4);
     for(int i = width; i > 0; i -= 16) {
-        canvas_draw_line(canvas, i, SHAPSHUP_TOP_SCALE, i, SHAPSHUP_TOP_SCALE + 4);
-        canvas_draw_line(canvas, i - 5, SHAPSHUP_TOP_SCALE, i - 5, SHAPSHUP_TOP_SCALE + 2);
-        canvas_draw_line(canvas, i - 10, SHAPSHUP_TOP_SCALE, i - 10, SHAPSHUP_TOP_SCALE + 2);
+    canvas_draw_line(canvas, i, SHAPSHUP_TOP_SCALE, i, SHAPSHUP_TOP_SCALE + 4);
+    canvas_draw_line(canvas, i - 5, SHAPSHUP_TOP_SCALE, i - 5, SHAPSHUP_TOP_SCALE + 2);
+    canvas_draw_line(canvas, i - 10, SHAPSHUP_TOP_SCALE, i - 10, SHAPSHUP_TOP_SCALE + 2);
     }
 
     canvas_draw_line(canvas, 0, SHAPSHUP_TOP_SCALE, width, SHAPSHUP_TOP_SCALE);
@@ -642,11 +642,16 @@ void format_number(uint64_t number_to_format, char* output_buffer) {
     snprintf(buf, sizeof(buf), "%lld", number_to_format);
     c = 2 - (int32_t)(strlen(buf) % 3);
     for(p = buf; *p != 0; p++) {
-        *output_buffer++ = *p;
-        if(c == 1) {
-            *output_buffer++ = ',';
-        }
-        c = (c + 1) % 3;
+    *output_buffer++ = *p;
+    if(c == 1) {
+        *output_buffer++ = ',';
+    }
+    c = (c + 1) % 3;
     }
     *--output_buffer = 0;
+}
+
+void format_frequency(uint32_t frequency, char* output_buffer) {
+    snprintf(
+        output_buffer, 32, "%03ld.%03ld MHz", frequency / 1000000 % 1000, frequency / 1000 % 1000);
 }

--- a/views/shapshup_main_view.c
+++ b/views/shapshup_main_view.c
@@ -481,7 +481,7 @@ void shapshup_main_view_draw_scale(Canvas* canvas, ShapShupMainViewModel* model)
     char buffer[64];
 
     canvas_set_color(canvas, ColorBlack);
-    canvas_set_font(canvas, FontBatteryPercent);
+    canvas_set_font(canvas, FontSecondary);
     uint8_t width = canvas_width(canvas);
 
     // First and last scale pin

--- a/views/shapshup_main_view_private.h
+++ b/views/shapshup_main_view_private.h
@@ -17,6 +17,8 @@
 #define SHAPSHUP_CHART_LOWEST_POINT 40
 #define SHAPSHUP_CHART_HIGHEST_POINT 20
 #define SCREEN_WIDTH 128
+#define SHAPSHUP_ALERT_DEFAULT_TTL 150
+#define SHAPSHUP_SCALE_MIN_VALUE 0.1
 
 /**
  * @brief ShapShupShapeItem
@@ -32,8 +34,16 @@ typedef struct {
     uint8_t height;
 } ShapShupShapeItem;
 
+/**
+ * @brief Construct a new array def object
+ * 
+ */
 ARRAY_DEF(ShapShupShapeItemArray, ShapShupShapeItem, M_POD_OPLIST)
 
+/**
+ * @brief init array
+ * 
+ */
 #define M_OPL_ShapShupShapeItemArray_t() ARRAY_OPLIST(ShapShupShapeItemArray, M_POD_OPLIST)
 
 /**
@@ -96,8 +106,28 @@ typedef struct {
     char alert_text[SHAPSHUP_TEXT_STORE_SIZE];
 } ShapShupMainViewModel;
 
+/**
+ * @brief shapshup_main_view_show_alert
+ * 
+ * @param instance ShapShupMainView*
+ * @param text to show
+ * @param ttl TTL value
+ */
 void shapshup_main_view_show_alert(ShapShupMainView* instance, const char* text, uint32_t ttl);
+
+/**
+ * @brief shapshup_main_view_reset_alert
+ * 
+ * @param instance ShapShupMainView*
+ */
 void shapshup_main_view_reset_alert(ShapShupMainView* instance);
+
+/**
+ * @brief shapshup_main_view_draw_alert_if_needed
+ * 
+ * @param canvas 
+ * @param model 
+ */
 void shapshup_main_view_draw_alert_if_needed(Canvas* canvas, ShapShupMainViewModel* model);
 
 /**
@@ -129,10 +159,18 @@ uint64_t calc_offset_per_page(uint64_t total, uint64_t min_len, float scale);
 /**
  * @brief format_number
  * 
- * @param n 
- * @param out 
+ * @param n numbver
+ * @param out output value
  */
 void format_number(uint64_t n, char* out);
+
+/**
+ * @brief format frequency to show
+ * 
+ * @param frequency frequency
+ * @param output_buffer output value
+ */
+void format_frequency(uint32_t frequency, char* output_buffer);
 
 /**
  * @brief shapshup_main_view_input

